### ISSUE TITLE
Exoplayer Download & Play Simultaneously

### ISSUE THREAD

**Issue by asaadco** [google/ExoPlayer#7311]
### [REQUIRED] Searched documentation and issues
I have looked into Stackoverflow, ExoPlayer documentation and this repository for any similar issues but I didn't find any results. 

### [REQUIRED] Question
I would like ExoPlayer to download a few, not necessarily in order, DASH segments through HTTP and read the rest locally from Cache folder. The goal here is to enable two people (Alice & Bob) to watch a video stream collaboratively.

For example, Alice downloads the odd-numbered DASH segments and shares it locally with Bob who is downloading the even-numbered segments and sharing it locally too with Alice. (If Alice or Bob fails to DL a segment on time, they download it themselves through HTTP which acts as the fallback).

What parts of exoplayer should I be looking at and modify? I was thinking of defining a DownloadManager to download\maintain a cache folder for each Alice and Bob (on one thread). The video stream will use the cache folder as the main media source and fall back to HTTP if needed (on another thread). The logic is that, before ExoPlayer decides to download a segment it checks the cache folder first. Please advise! Thank you.

**Comment by marcbaechinger** [comment]
Are Alice and Bob each having a dedicated device, app, player or is everything in a single player instance? Are they watching simultaneously or sequentially?

If it's only about a single instance of ExoPlayer, then you can just go the route af a single Cache pointing to the same cache directory. That should work. I don't think there is a difference between Alice/Bob and Bob/Bob in this case.

**Comment by asaadco** [comment]
Thank you for the quick response, Marc. :)

Before the start of the stream, Alice and Bob will know what segments (game plan) they have to download through HTTP and what segments they will have to wait to receive through p2p. They watch the video simultaneously (this can be relaxed though). Alice could be downloading the first 20 segments while Bob downloads only the last 10 segments of the video. 

Each have their own dedicated device, app, player and bandwidth. The only communication medium between Alice and Bob is WiFi-Direct. Once Alice downloads a segment through HTTP, she will forward it to Bob immediately through WiFi-Direct and vice versa.

**Comment by marcbaechinger** [comment]
It sounds what you would need is something like a distributed Cache, or a Cache which syncs between devices. This sounds like a complicated thing to do in general. Plus, with phones you need to make sure that they have a client/server or p2p infrastructure of some kind to share that data and coordinate this with the original upstream connection to the server. All of these things are not supported by ExoPlayer out of the box. It needs you significant effort to get there I think.

From the point of view of the system architecture that you want to establish between a remote server with the media and two phones, I'd actually recommend to solve this with an HTTP proxy which is between the phones and the server. So instead of getting the media directly from the remote server, the requests go through a HTTP proxy. The HTTP proxy would need to support range requests in some ways I think, but that's probably doable as proxies (i.e. Apache with a certain module) are quite sophisticated.

If you want to code this from scratch on the client side in the app that's probably a difficult endeavour. It would need you to implementat of a `Cache` which can sync between devices.

**Comment by asaadco** [comment]
That makes sense. Would it possible to modify ExoPlayer behavior to skip downloading certain segments and persist when there are missing segments (Like what happens in live streams)? I have looked into solutions such as CDNBye which uses WebRTC as a medium and an HTTP proxy like you said but I would like to experiment with implementing this on the client-side first though. What ExoPlayer modules should I be looking at to influence the order of segments download? I tried to look into DashMediaSource and DashDownloader but I don't think they have what I want.

Again, thank you so much!
